CONTRIBUTING.md

```diff
@@ -53,9 +53,9 @@ In addition to writing properly formatted commit messages, it's important to inc
 
 Package version upgrades usually allow for simpler commit messages, including attribute name, old and new version, as well as a reference to the relevant release notes/changelog. Every once in a while a package upgrade requires more extensive changes, and that subsequently warrants a more verbose message.
 
-We prefer not to use the "squash merge" feature in nixpkgs: in order to keep as much information as possible in the commit history, we expect pull requests to consist of self-contained commits as described above.
-This means that, after addressing review comments and before the PR is merged, you will sometimes need to rewrite your branch's history and then force-push it with `git push --force-with-lease`.
-Useful commands to be comfortable with are `git commit --amend`, `git commit --fixup` and `git rebase -i` (and don't forget that git lets you define aliases!).
+Pull requests should not be squash merged in order to keep complete commit messages and GPG signatures intact and must not be when the change doesn't make sense as a single commit.
+This means that, when addressing review comments in order to keep the pull request in an always mergeable status, you will sometimes need to rewrite your branch's history and then force-push it with `git push --force-with-lease`.
+Useful git commands that can help a lot with this are `git commit --patch --amend` and `git rebase --interactive @~3`. For more details consult the git man pages.
```

Comment:
FYI: I would've omitted the `@~3` from `git rebase --interactive` as that causes even more trouble if users execute that command without understanding it (e.g., they will duplicate two commits if they execute it on their PR branch that only contains a single additional commit - the main problem being that some Nixpkgs maintainers might not even notice it and we'll have duplicate commits in the history after the merge).

(Using something like `git merge-base --fork-point origin/master` could be a bit better but also has lot's of issues as it too makes lots of assumptions (`origin/master`) so it doesn't really make sense either.)

It would probably be much better to link to some external documentation like https://git-rebase.io/ or https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History instead.

PS: You discarded the commit message body when you used GHs squash merge functionality to merge this PR (which is both ironic and undesirable as https://github.com/NixOS/nixpkgs/commit/756a5e6b74f786f08fa3a4c006b31fa316133566 now lacks any reasoning for the change... :o).

PPS: +1 for the `complete commit messages and GPG signatures` part

Reply:
I like the idea of omitting `@~3` and linking each git command to its manual (or third-party documentation).

Reply:
#207487